core/core-impl/src/main/java/ai/timefold/solver/core/impl/util/ConstantLambdaUtils.java

```diff
@@ -0,0 +1,59 @@
+package ai.timefold.solver.core.impl.util;
+
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.ToIntBiFunction;
+import java.util.function.ToIntFunction;
+
+import ai.timefold.solver.core.api.function.QuadFunction;
+import ai.timefold.solver.core.api.function.ToIntQuadFunction;
+import ai.timefold.solver.core.api.function.ToIntTriFunction;
+import ai.timefold.solver.core.api.function.TriFunction;
+
+public final class ConstantLambdaUtils {
```

Comment:
If these functions are returned via static methods, you can get the casting done inside the methods, as opposed to plaguing the entire codebase with it.

Example: 

    public static final <A, B> BiPredicate<A, B> notEquals() {
        return (BiPredicate<A, B>) NOT_EQUALS; 
    }

Reply:
Also, let's add a Javadoc to the class, explaining why we need to do something like this. Every time there is such non-standard behavior, there should be an explanation as to why it needs to be like that.